### ISSUE TITLE
Improve docs explaining commands possible inputs

### ISSUE THREAD

**Issue by DavidGOrtega** [iterative/cml#361]
Coming from [discord](https://discord.com/channels/485586884165107732/728693131557732403/784442370497970178)

Users might not even know what are the posible inputs that the commands have

**Comment by elleobrien** [comment]
Definitely. Our docs are un-ideally formatted for explaining the inputs and flags that `cml-publish` takes... let me think on how we can improve beyond our current table of functions, which is a) hard to read for arguments/flags, and b) not comprehensive.

**Comment by elleobrien** [comment]
Two options come to mind:

1. Add a new section to the Readme, like "Advanced Function Docs", where we list every flag/argument for `cml-publish`, `cml-send-comment` and `cml-tensorboard-dev` (and soon `cml-cloud-runner`).

2. It's time to get real paginated docs, like DVC has. 

I think we're almost ready for option 2, because cloud runner is about to introduce many more flags/options and multiple providers. As one big Readme, this seems unsustainable. 

Thoughts?

**Comment by 0x2b3bfa0** [comment]
I think that it would be also a good idea to document the command-line interface as thoroughly as possible at the code level and then generate the documentation automatically from there. [Typer](https://typer.tiangolo.com/typer-cli/#generate-docs) offers a good example of this approach.

It looks like [`oclif`](https://github.com/oclif/oclif#-features) does it automatically by design and [`yargs`](https://github.com/EliseevNP/cli-docs-generator) also supports it to some extent through third-party packages.

**Comment by jorgeorpinel** [comment]
Hi 👋 

> document the command-line interface as thoroughly as possible at the code level and then generate the documentation automatically from there

On this I can tell you at least on the DVC side so far we have avoided this strategy because

1. The core team has a "self-documenting code" policy so comments and even docstrings tend to be avoided except for certain parts of the codebase that we may expect users to look at (internal `Repo` API and public `dvc.api`).
2. There doesn't seem to be an easy way to generate docs with the final formatting even with flexible tools we want (e.g. Sphynx in the case of Python) so it will still require significant manual edition.

That said if you deem useful/worth-it to maintain structured code documentation in the CML codebase, whatever can get auto-generated from there would probably help a lot as an intermediate step (could even wrap is into a DVC pipeline haha).

**Comment by 0x2b3bfa0** [comment]
👋  Hi, @jorgeorpinel!

1. I wasn't familiar with the DVC self-documenting code policy, but it would be great to keep CML as aligned as possible with its sibling projects. I personally like that idea, at least for the most expressive languages and codebases.

2. I'm [somewhat accustomed](https://0x2b3bfa0.github.io/python-blobopera/blobopera.recording.html) to fully automated documentation generation with [Napoleon](https://sphinxcontrib-napoleon.readthedocs.io) and the results aren't that bad for small projects, but I can't think of any practical way of doing something similar in this case.

Regarding command-line parameter documentation, I don't know if (semi-)automating it could be worth the effort, but it might be an interesting to consider it in the future. For now, keeping manually written documentation it two places isn't that bad.

_Post scriptum: it would be funny to use DVC as part of the documentation process._ 😂

**Comment by jorgeorpinel** [comment]
> wasn't familiar with the DVC self-documenting code policy

On that I would refer you to @efiop and @shcheklein mainly 🙂

> keeping manually written documentation it two places

What do you mean? (What are the 2 places)? Thanks

**Comment by 0x2b3bfa0** [comment]
> _What do you mean? (What are the 2 places)?_

The number actually varies if you count the code itself. 😅  We have the `cml-runner` command line argument reference with slightly different description in a few different places:

* [The documentation page](https://github.com/iterative/dvc.org/blob/master/content/docs/cml/self-hosted-runners.md#arguments)
* [The repository itself](https://github.com/iterative/cml#arguments)
* [The code](https://github.com/iterative/cml/blob/81e8a363eb300306c888a77bdeb4d71179f09203/bin/cml-publish.js#L39-L82)

**Comment by jorgeorpinel** [comment]
I see. Yeah probably no need to duplicate all the details in the README since there are docs online. But having a summary of all the basics in the README is great for people who expect that in GH (linking to the docs for more details).

**Comment by 0x2b3bfa0** [comment]
Referencing iterative/dvc.org#2443...

**Comment by 0x2b3bfa0** [comment]
* Linking https://github.com/yargs/yargs/issues/774 and https://www.npmjs.com/package/cli-docs-generator/v/1.0.1
 for future references.
